### PR TITLE
Throw exception when incompatible settings are configured

### PR DESCRIPTION
**Pull request by lailabougria**

### CODE REVIEW

File: src/NServiceBus.Persistence.AzureTable/SagaPersisters/CompatibilitySettings.cs [hunk 1]
```diff
@@ -40,8 +49,17 @@ public void AllowSecondaryKeyLookupToFallbackToFullTableScan()
...
+                throw new InvalidOperationException("Compatibility mode was disabled. AssumeSecondaryKeyUsesANonEmptyRowKeySetToThePartitionKey requires compatibility mode to be enabled.");
+
             this.GetSettings().Set(WellKnownConfigurationKeys.SagaStorageCompatibilityMode, true);
             this.GetSettings().Set(WellKnownConfigurationKeys.SagaStorageAssumeSecondaryKeyUsesANonEmptyRowKeySetToThePartitionKey, true);
+            wasAssumeSecondaryKeyUsesANonEmptyRowKeySetToThePartitionKeySet = true;
         }
+        
+        bool wasCompatibilityModeDisabled { get; set; }
+        bool wasAssumeSecondaryKeyUsesANonEmptyRowKeySetToThePartitionKeySet { get; set; }
+        bool wasAssumeSecondaryIndicesExistSet { get; set; }
```
Comment (SeanFeldman) [inline src/NServiceBus.Persistence.AzureTable/SagaPersisters/CompatibilitySettings.cs:62]: `wasAssumeSecondaryIndicesExistSet` --> assumeSecondaryIndexesExistWasSet

File: src/NServiceBus.Persistence.AzureTable/SagaPersisters/CompatibilitySettings.cs [hunk 2]
```diff
@@ -40,8 +49,17 @@ public void AllowSecondaryKeyLookupToFallbackToFullTableScan()
...
+            if (wasCompatibilityModeDisabled)
+                throw new InvalidOperationException("Compatibility mode was disabled. AssumeSecondaryKeyUsesANonEmptyRowKeySetToThePartitionKey requires compatibility mode to be enabled.");
+
             this.GetSettings().Set(WellKnownConfigurationKeys.SagaStorageCompatibilityMode, true);
             this.GetSettings().Set(WellKnownConfigurationKeys.SagaStorageAssumeSecondaryKeyUsesANonEmptyRowKeySetToThePartitionKey, true);
+            wasAssumeSecondaryKeyUsesANonEmptyRowKeySetToThePartitionKeySet = true;
         }
+        
+        bool wasCompatibilityModeDisabled { get; set; }
+        bool wasAssumeSecondaryKeyUsesANonEmptyRowKeySetToThePartitionKeySet { get; set; }
```
Comment (SeanFeldman) [inline src/NServiceBus.Persistence.AzureTable/SagaPersisters/CompatibilitySettings.cs:61]: `wasAssumeSecondaryKeyUsesANonEmptyRowKeySetToThePartitionKeySet` --> assumeSecondaryKeyUsesANonEmptyRowKeySetToThePartitionKeyWasSet

File: src/NServiceBus.Persistence.AzureTable/SagaPersisters/CompatibilitySettings.cs [hunk 3]
```diff
@@ -40,8 +49,17 @@ public void AllowSecondaryKeyLookupToFallbackToFullTableScan()
...
         {
+            if (wasCompatibilityModeDisabled)
+                throw new InvalidOperationException("Compatibility mode was disabled. AssumeSecondaryKeyUsesANonEmptyRowKeySetToThePartitionKey requires compatibility mode to be enabled.");
+
             this.GetSettings().Set(WellKnownConfigurationKeys.SagaStorageCompatibilityMode, true);
             this.GetSettings().Set(WellKnownConfigurationKeys.SagaStorageAssumeSecondaryKeyUsesANonEmptyRowKeySetToThePartitionKey, true);
+            wasAssumeSecondaryKeyUsesANonEmptyRowKeySetToThePartitionKeySet = true;
         }
+        
+        bool wasCompatibilityModeDisabled { get; set; }
```
Comment (SeanFeldman) [inline src/NServiceBus.Persistence.AzureTable/SagaPersisters/CompatibilitySettings.cs:60]: `wasCompatibilityModeDisabled` --> compatibilityModeDisabledWasSet

File: src/NServiceBus.Persistence.AzureTable/SagaPersisters/CompatibilitySettings.cs [hunk 4]
```diff
@@ -20,7 +22,10 @@ internal CompatibilitySettings(SettingsHolder settings) : base(settings)
         /// </summary>
         public void DisableSecondaryKeyLookupForSagasCorrelatedByProperties()
         {
+            if (wasAssumeSecondaryIndicesExistSet || wasAssumeSecondaryKeyUsesANonEmptyRowKeySetToThePartitionKeySet)
+                throw new InvalidOperationException("Compatibility mode cannot be disabled when AllowSecondaryKeyLookupToFallbackToFullTableScan or AssumeSecondaryKeyUsesANonEmptyRowKeySetToThePartitionKey are called.");
             this.GetSettings().Set(WellKnownConfigurationKeys.SagaStorageCompatibilityMode, false);
+            wasCompatibilityModeDisabled = true;
```
Comment (danielmarbach) [inline src/NServiceBus.Persistence.AzureTable/SagaPersisters/CompatibilitySettings.cs:28]: I don't think this is going to work. All the state is in settings and it is perfectly valid for someone to have configuration like

```
persistence.Compatibility().Method();
//somewhere else
persistence.Compatibility().OtherMethod();
```

you need to reach into settings to actually validate the state

File: src/NServiceBus.Persistence.AzureTable/SagaPersisters/CompatibilitySettings.cs [hunk 5]
```diff
@@ -1,3 +1,5 @@
+using System;
```
Comment (danielmarbach) [inline src/NServiceBus.Persistence.AzureTable/SagaPersisters/CompatibilitySettings.cs:1]: The lovely using ;)

File: src/NServiceBus.Persistence.AzureTable/SagaPersisters/CompatibilitySettings.cs (hunk 1 above)
Comment (lailabougria) [reply src/NServiceBus.Persistence.AzureTable/SagaPersisters/CompatibilitySettings.cs:62]: That's how I had it, and really find was*propertyname*set a more readable variant 😬

Comment (lailabougria) [reply src/NServiceBus.Persistence.AzureTable/SagaPersisters/CompatibilitySettings.cs:62]: I'll leave it to @danielmarbach to be the tiebreaker 😝

**Review by danielmarbach — APPROVED** [review]

**Pull request merged**